x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/common/CartesianPoint.java

```diff
@@ -157,124 +156,33 @@ public String toString() {
-                            }
-                        } else {
-                            throw new ElasticsearchParseException(
-                                "field must be either [{}] or [{}]",
-                                X_FIELD.getPreferredName(),
-                                Y_FIELD.getPreferredName()
-                            );
-                        }
-                    } else {
-                        throw new ElasticsearchParseException("token [{}] not allowed", subParser.currentToken());
-                    }
-                }
-            }
-            if (numberFormatException != null) {
-                throw new ElasticsearchParseException(
-                    "[{}] and [{}] must be valid double values",
-                    numberFormatException,
-                    X_FIELD.getPreferredName(),
-                    Y_FIELD.getPreferredName()
-                );
-            } else if (Double.isNaN(x)) {
-                throw new ElasticsearchParseException("field [{}] missing", X_FIELD.getPreferredName());
-            } else if (Double.isNaN(y)) {
-                throw new ElasticsearchParseException("field [{}] missing", Y_FIELD.getPreferredName());
-            } else {
-                return point.reset(x, y);
-            }
-
-        } else if (parser.currentToken() == XContentParser.Token.START_ARRAY) {
-            try (XContentSubParser subParser = new XContentSubParser(parser)) {
-                int element = 0;
-                while (subParser.nextToken() != XContentParser.Token.END_ARRAY) {
-                    if (subParser.currentToken() == XContentParser.Token.VALUE_NUMBER) {
-                        element++;
-                        if (element == 1) {
-                            x = subParser.doubleValue();
-                        } else if (element == 2) {
-                            y = subParser.doubleValue();
-                        } else {
-                            throw new ElasticsearchParseException(
-                                "[{}}] field type does not accept > 2 dimensions",
-                                PointFieldMapper.CONTENT_TYPE
-                            );
-                        }
-                    } else {
-                        throw new ElasticsearchParseException("numeric value expected");
-                    }
-                }
-            }
-            return point.reset(x, y);
-        } else if (parser.currentToken() == XContentParser.Token.VALUE_STRING) {
-            String val = parser.text();
-            return point.resetFromString(val, ignoreZvalue);
-        } else {
-            throw new ElasticsearchParseException("point expected");
-        }
+        return cartesianPointParser.parsePoint(parser, ignoreZValue, value -> {
+            CartesianPoint point = new CartesianPoint();
+            point.resetFromString(value, ignoreZValue);
+            return point;
+        }, value -> null);
```

Comment:
Instead of null, Should be throw an IllegalArgumentException here?

Reply:
It was already throwing an exception. The list of valid fields is checked much earlier. I've added a number of tests to assert on the actual exception that should be thrown for all the possible invalid formats. One thing the new test shows is that seemingly similar things (like test for NaN) are tested quite differently (and produce different error messages) for different code paths (geojson parser, compared to string parser, compared to array parser). I've not attempted to unify the error messages, for backwards compatibility, but just assert that the expected errors are thrown.